src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -3,69 +3,167 @@
+    /// in-memory buffer, then write that buffer's length to the pipe-writer, then copy the in-memory buffer to the
+    /// writer.
+    /// </para>
+    /// When writing/reading the data-segment, we use an the <see cref="ObjectWriter"/>/<see cref="ObjectReader"/>
+    /// subsystem.  This will write its own validation bits, and then the data describing the asset.  This data is:
+    /// <code>
+    /// -----------------------------------------------------------------------------------------------------------
+    /// | data (variable length)                                                                                  |
+    /// -----------------------------------------------------------------------------------------------------------
+    /// | ObjectWriter validation (2 bytes) | checksum (16 bytes) | kind (4 bytes) | asset-data (asset specified) |
+    /// -----------------------------------------------------------------------------------------------------------
+    /// </code>
+    /// The validation bytes are followed by the checksum.  This is needed in the message as assets can be found in any
+    /// order (they are not reported in the order of the array of checksums passed into the writing method). Following
+    /// this is the kind of the asset.  This is used by the reading code to know which asset-deserialization routine to
+    /// invoke. Finally, the asset data itself is written out.
+    /// </summary>
     internal static class RemoteHostAssetSerialization
     {
+        /// <summary>
+        /// A sentinel byte we place between messages.  Ensures we can detect when something has gone wrong as soon as possible.
+        /// </summary>
+        private const byte s_messageSentinelByte = 0b01010101;
+
+        private static readonly ObjectPool<SerializableBytes.ReadWriteStream> s_streamPool = new(SerializableBytes.CreateWritableStream);
+
         public static async ValueTask WriteDataAsync(
-            Stream stream,
-            Dictionary<Checksum, object> assetMap,
-            ISerializerService serializer,
-            SolutionReplicationContext context,
-            Checksum solutionChecksum,
+            PipeWriter pipeWriter,
+            AssetPath assetPath,
             ReadOnlyMemory<Checksum> checksums,
+            SolutionAssetStorage.Scope scope,
+            ISerializerService serializer,
             CancellationToken cancellationToken)
         {
-            using var writer = new ObjectWriter(stream, leaveOpen: true, cancellationToken);
+            var foundChecksumCount = 0;
 
-            // This information is not actually needed on the receiving end.  However, we still send it so that the
-            // receiver can assert that both sides are talking about the same solution snapshot and no weird invariant
-            // breaks have occurred.
-            solutionChecksum.WriteTo(writer);
+            await scope.AddAssetsAsync(
```

Comment:
note: one thing we could consider is using an ABWQ, where we pass a callback that just adds the checksums/assets to the ABWQ.  The ABQG would then write the results to the pipe-writer.  The benefit of that is less async calls in AddAssetsAsync, and the ability for ABWQ to just have it's own loop on its own thread where it does the writing in bulk (still an asset at a time, so the reading end can read immediately).

If we do this, we would be able to then have AddAssetsAsync process the solutoin in parallel, which could speed up searches signficantly.  

(we'd also need to pass a concurrentset to the AddAssetsAsync to make things safe).